Add: Autostart

rtmiz opened the pull request:
Consolidation of Build and Start buttons in the custom image building UI into one button.

Permalinks now feature a autoStart field, that when set to true will instantly initiate the spawning process of the given environment.

rtmiz committed:
Add: Autostart

batpad commented:
@GeraldIr this looks good to me!

Do you mind writing up a couple sentences on what this solves and exactly what the expected behaviour here would be? And then am happy to test.

It would also be nice to figure out writing a test for this - I can help with that once we have this outlined a bit.

Thank you for your contribution!

GeraldIr commented:
I've found the workflow of Build Image -> Start Image a bit cumbersome, so I was just looking for a way to simplify it down to a single press of a button.

Ultimately I'd like to add an auto-start option for permalinks as well, in order to go from Click Permalink -> Running JupyterLab in a single click from anywhere without further user-input.

This is a feature for the original BinderHub UI so that's what I'm trying to replicate.

To be honest, I mistakenly opened the PR early, so for now I'll change this to a draft.

rtmiz committed:
Add: auto-start via permalink

GeraldIr commented:
As for the security concerns, as the standalone binderhub UI already works the same way I don't see much of an issue with this;

but I am not a cybersecurity expert and usually take such concerns too lightly, so there is that 😅

rtmiz committed:
Updates: Consolidate build and start button into one, add autoStart: false to copied permalink to make creating auto-start urls easier

rtmiz commented:
I've now consolidated the Build and Start buttons from the custom image builder into a single button, as well as made it easier to create auto-starting permalinks, by including the autoStart option set to false by default when copying a permalink.

batpad commented:
> As for the auto-start feature, this is something we were looking for in our deployment and have now used to great success. We combined the autoStart option with multiple ?next= referrals (for login as well as invoking nbgitpuller) to cut down the workflow for users. This basically allowed us to generate URLs which are a 1-click referral to a specific environment with a specific notebook and it's been a great boon for tutorial sessions and our notebook gallery. Copied permalinks also now include a "autoStart": "false" by default, to make generating autoStart URLs easier.

That sounds super-cool! If you have anything public that you're able to share, I'd love to get a better sense of how y'all are using this.

rtmiz commented:
@batpad You can find our notebook gallery here: https://eopf-sample-service.github.io/eopf-sample-notebooks/gallery/

From there select a notebook and click the Launch in JupyterHub button! You'll need to create an account though to use it, but otherwise it should be accessible.

Also, there seems to still be some issues related to permalinks (values don't get updated if you opened the page via one and autoStart doesn't work yet for the build image page). I'm working on these right now.

I can also share a piece of code when these issues are fixed that generates these URLs, this would also be of interest to @maxrjones for his second issue #137

rtmiz committed:
Add: timeout when autoStarting so the form can properly load

rtmiz commented:
To fix the issue I added a timeout when auto-starting so the form can properly load (this is a hack imo, but I don't know enough about react to fix this in a proper manner).

@maxrjones Here's the code I mentioned earlier, make sure the image you are building will have nbgitpuller installed:

```
import urllib.parse
import re

# Change these values

# full url from github to the relevant ipynb, e.g.: https://github.com/org/repo/blob/branch/notebooks/example.ipynb
notebook_url = "https://github.com/org/repo/blob/branch/notebooks/example.ipynb"

# permalink copied from filled out fancy profiles form, this can be from the build your own image UI
permalink = ""

# JupyterHub URL e.g.: https://hub.example.org
jupyterhub_url = "https://hub.example.org

# URL creation
notebook_suburl = notebook_url.split("/blob/")[0].split("/")[-1] + "/" + "/".join(notebook_url.split("/blob/")[1].split("/")[1:])
branch = notebook_url.split("/blob/")[1].split("/")[0]
git_repo_url = notebook_url.split("/blob/")[0]
gitpull_next = f"/hub/user-redirect/git-pull?repo={git_repo_url}&branch={branch}&urlpath=lab/tree/{notebook_suburl}"
spawn_next = '/hub/spawn?next=' + urllib.parse.quote(gitpull_next, safe='')
occ = re.search("%23.+?(?=%7D)", permalink)
fancy_forms_config = occ[0].replace("%22autoStart%22%3A%22false%22", "%22autoStart%22%3A%22true%22") # Make sure autoStart is set to true (you can comment this out if it's not important)
finished_url = f'{jupyterhub_url}/hub/login?next=' + urllib.parse.quote(spawn_next, safe='') + fancy_forms_config + "%7D"

print(finished_url)
```

rtmiz committed:
Fix: tests, linting. Wrap ProfileForm rendering in form tag for consistency.

rtmiz committed:
Fix: further linting issues

rtmiz commented:
@batpad No worries!

batpad commented on src/ProfileForm.tsx:
```diff
@@ -104,6 +131,22 @@ function Form() {
+      if (form) {
+        const button = form.querySelector("button[type=\"submit\"]") as HTMLButtonElement | null;
+        if (button) {
+          setTimeout(() => {
```
This seems like a neat hack indeed. Would just be curious to see if there's some-how a better way to do this, this does seem like it could get brittle and hard to debug in some edge case scenarios.

hanbyul-here replied:
Echoing  @batpad 's concern - there are two ways of submitting the form in this file. Any reason of not using `requestSubmit` (https://github.com/2i2c-org/jupyterhub-fancy-profiles/pull/132/files) for this case?

batpad commented on src/ImageBuilder.test.tsx:
```diff
@@ -8,7 +8,11 @@ import renderWithContext from "./test/renderWithContext";
 
-  renderWithContext(<ProfileForm />);
+  renderWithContext(
+    <form>
```
Needing to wrap the `<form>` around the `<ProfileForm>` in the tests makes sense to me - since fancy-profiles gets "injected" onto a page that already contains the container `<form>` tag, the `<ProfileForm>` expects to be rendered inside an existing `<form>` - since that does not exist during tests, we add it this way.

This looks a little awkward, but makes sense to me. Just giving @hanbyul-here some context while reviewing - not sure if there's a way to do this that feels more elegant, but this seems okay to me.

hanbyul-here replied:
🤔 imo, it is ideal if the test's scope is narrowed enough that the test doesn't have to cover any unnecessary tasks. Most tests were scoped down well enough that it doesn't need additional wrapper. I understand some tests might need this wrapper now since this PR makes form submission this application's responsibility. but I still feel like we can get rid of this additional wrapper for the ones that don't need one? 

I think one of the biggest change this PR introduces is the programatic form submission - 
There’s something nice about letting HTML handle form submission — it keeps the application’s responsibility focused on processing and cleaning the data. This PR expands that responsibility. That’s not necessarily a bad thing, but we would at least need a test for it. 
Maybe we can add an explicit test to check if form submission worked with this additional wrapper?

rtmiz committed:
Add: Documentation for Auto-Start feature, nbgitpuller integration for full auto-start (login->profile-selection->notebook) configuration

rtmiz committed:
Add: reference to auto start and permalinks in the README.md

batpad committed:
Merge branch 'main' into auto-start

hanbyul-here commented on src/ProfileForm.tsx:
```diff
@@ -194,7 +237,7 @@ function Form() {
         onClick={handleSubmit}
       >
-        Start
+        { buildImageStart ? "Build Image and Start" : "Start" }
```
This doesn't get synced with the user's current choice - When I toggle back to one of the default options, the button still says 'Build Image and Start'  (and will start the server with custom image instead of one of the default options that I selected) 

 If I am reading the flow correctly, there is no way to prevent a user from hitting submit multiple times while the image is building, and there is no 'check' to see if the image building is already in progress. I am concerned that this will result in multiple build triggers.

rtmiz replied:
Unsure what you mean by this not syncing with the user's current choice? This is how it looks on my end.

![output](https://github.com/user-attachments/assets/5d68c967-73f1-4c46-9840-26e45cc4681b)

As for the multi-submit, you are right, there is currently nothing stopping the user from either submitting the form again or changing to a different drop-down item during the build process. I'll see to it that the form gets disabled during building.

hanbyul-here replied:
re:Syncing user's choice: Ah! I forgot that dev environment is different per config. My config (default config that comes with the app) has more than one profile, so user can navigate between profiles.  (There is a case where ImageBuilder is not unmounted yet, while a user's current choice is not building a custom image, which leaves the text on the button unsync) Is this very unlikely configuration set up in real life? 

https://github.com/user-attachments/assets/e3b17fd0-fa9c-460c-b445-2147334431f7

rtmiz replied:
Ahh, I see, I can replicate this on one of my deployments! I'll see if I can find a fix for this.

rtmiz replied:
This should be fixed with my newest changes.

hanbyul-here commented on src/ProfileForm.tsx:
```diff
@@ -30,16 +31,43 @@ function Form() {
+    if (buildImageStart){
+      await buildImageStart();
+    }
+    setShouldSubmit(true);
```
I might be missing something but I can't figure out why this needs to be handled indirectly. Can we directly call `form.requestSubmit`here?

rtmiz replied:
Nope, you're right I've changed both this and the other occurrence to form.requestSubmit()

rtmiz replied:
I actually take this back, I didn't test it properly, using form.requestSubmit in either of these circumstances does not work for reasons unknown to me (but presumably the reason I have it in there how it is, it's been a while now).

I agree that there should be a neater way of doing all of this, but I've ended up with what works over what seems like it should work here

hanbyul-here commented on src/ImageBuilder.tsx:
```diff
@@ -196,30 +218,41 @@ export function ImageBuilder({ name, isActive, optionKey }: ICustomOptionProps)
+    return () => {
+      setBuildImageStart(null);
+    };
+  }, []);
```
I am guessing a lot of added `useRefs` are to get updated values for `handleBuildStart`. I wonder if we can trigger the this hook with those values that need to be updated? (moving dependencies of this hook https://github.com/2i2c-org/jupyterhub-fancy-profiles/pull/132/files#diff-fa607bb96a32d3e9d15c039c5b36fa3806089f0a0023d231194f996ec3fea311R205 to here?)

hanbyul-here commented on src/ProfileForm.tsx:
```diff
@@ -30,16 +31,43 @@ function Form() {
+    if (shouldSubmit) {
+      submitTheForm();
+      setShouldSubmit(false);
+      const form = document.querySelector("form");
```
This comment is not only for this line, butI noticed that `querySelector` is used over `.closest` in this PR. Even though it is quite unlikely situation that there will be more than one `<form />` on the page, using `.closest("form");` seems to make sense since it is reaching for the nearest form element, not the first form element? am I missing anything?

rtmiz replied:
afaik `.closest` is a method of Element, I don't have a handy element to call `.closest` on anywhere so I'm using `querySelector`.

hanbyul-here commented on src/ImageBuilder.tsx:
```diff
@@ -297,8 +320,7 @@ export function ImageBuilder({ name, isActive, optionKey }: ICustomOptionProps)
         aria-hidden="true"
         style={{ display: "none" }}
-        onInvalid={() =>
-          setCustomImageError("Wait for the image build to complete.")}
```
Shouldn't this message be displayed to indicate that image building is in progress?

hanbyul-here commented on src/ProfileForm.tsx:
```diff
@@ -30,16 +31,43 @@ function Form() {
+    const form = document.querySelector("form");
+
+
     const formIsValid = form.checkValidity();
```
Shouldn't form validation happen before building image?

rtmiz replied:
This is before building no? It's in the extra submitTheForm Function (arguably a bad function name), but we aren't building here yet, we're just doing manual form validation.

hanbyul-here replied:
Isn't the flow (from https://github.com/2i2c-org/jupyterhub-fancy-profiles/pull/132/files#diff-57e82f51ce2ff82498b68631c62c2bf5723ee306d5f89d5a4728012f0ae999e4R52-R56)  
buildImage if there is one => and display the form errors (`submitTheForm` but not blocking form submit)  => Submit ? Which means that even if Repository input is not valid, it will start the buildimage (even if it errors)

This behavior is also related to this comment https://github.com/2i2c-org/jupyterhub-fancy-profiles/pull/132/files#r2632003823

rtmiz replied:
I've reworked the flow now, if you could check it out again.

hanbyul-here reviewed (commented):
I am still getting myself familiar with the codebase but I thought it would be better to start the conversation early rather than hang up.  So please bear with me if I read anything wrong.

rtmiz committed:
Fix: failing tests, Update: use requestSubmit instead of varying other methods

rtmiz committed:
Revert: last commit requestSubmit changes, uppercase AS in Dockerfile

rtmiz committed:
Fix: autostart, Add: disable form during build

hanbyul-here commented on src/ProfileForm.tsx:
```diff
@@ -54,7 +82,6 @@ function Form() {
       }, 100);
 
       setProfileError(!selectedProfile ? "Select a container profile" : "");
-      e.preventDefault();
```
This line was preventing the invalid form from being submitted, but now the form can be submitted with invalid values.

hanbyul-here replied:
I am not sure if I missed something, but this is still not resolved as far as I can test. ex. I can put an invalid docker image (not binder image), but still can trigger the start of the server despite the error display. 

<img width="797" height="472" alt="Screenshot 2025-12-18 at 12 44 02 PM" src="https://github.com/user-attachments/assets/38db7304-018b-4ce3-aced-53876789e4a2" />

rtmiz replied:
This should be fixed with my newest update.

rtmiz committed:
Fix: state not syncing correctly, between different profiles

rtmiz committed:
Update: relocate validation logic to change flow
